dataplane/standalone/translator.cc

```diff
@@ -14,62 +14,91 @@
+  sai_status_t status = 0;
+  switch (type) {
+    case SAI_OBJECT_TYPE_SWITCH: {
+      auto sw = std::make_shared<Switch>(*id, this->attrMgr, this->client);
+      this->switches[*id] = sw;
+      this->apis[std::to_string(*id)] = sw;
+      status = this->apis[std::to_string(*id)]->create(attr_count, attr_list);
+      break;
+    }
+    default:
+      LOG(INFO) << "Unknown type " << type;
+      // TODO(dgrau): handle other types
+  }
 
-sai_status_t Translator::getAttribute(sai_object_id_t id,
-                                      sai_attribute_t* attr) {
-  auto iter = this->objects[id].attributes.find(attr->id);
-  if (iter == this->objects[id].attributes.end()) {
-    return SAI_STATUS_ITEM_NOT_FOUND;
+  if (status != SAI_STATUS_SUCCESS) {
+    return status;
   }
-  *attr = {
-      iter->first,
-      iter->second,
-  };
+  // Save attributes to attr managers.
+  for (uint32_t i = 0; i < attr_count; i++) {
+    this->attrMgr->set_attribute(std::to_string(*id), attr_list[i]);
+  }
+
   return SAI_STATUS_SUCCESS;
 }
 
+// create switch-scoped objects.
 sai_status_t Translator::create(sai_object_type_t type, sai_object_id_t* id,
-                                uint32_t attr_count,
+                                sai_object_id_t switch_id, uint32_t attr_count,
                                 const sai_attribute_t* attr_list) {
-  return sai_status_t();
+  *id = this->attrMgr->create(type, switch_id);  // Allocate new id for object.
+  auto status = this->switches[switch_id]->create_child(
+      type, *id, attr_count,
+      attr_list);  // Delegate creation to switch instance.
+  if (status != SAI_STATUS_SUCCESS) {
+    return status;
+  }
+  // Save attributes to attr managers.
+  for (uint32_t i = 0; i < attr_count; i++) {
+    this->attrMgr->set_attribute(std::to_string(*id), attr_list[i]);
+  }
+  return SAI_STATUS_SUCCESS;
 }
 
+// create entry type objects.
 sai_status_t Translator::create(sai_object_type_t type, common_entry_t id,
                                 uint32_t attr_count,
                                 const sai_attribute_t* attr_list) {
-  return sai_status_t();
-}
+  std::string idStr = this->attrMgr->serialize_entry(type, id);
+  sai_object_id_t switch_id = this->attrMgr->entry_to_switch_id(type, id);
 
-sai_status_t Translator::create(sai_object_type_t type, sai_object_id_t* id,
-                                sai_object_id_t switch_id, uint32_t attr_count,
-                                const sai_attribute_t* attr_list) {
-  return sai_status_t();
+  this->attrMgr->create(type, idStr, switch_id);
+
+  auto status = this->switches[switch_id]->create_child(
+      type, id, attr_count,
+      attr_list);  // Delegate creation to switch instance.
+  if (status != SAI_STATUS_SUCCESS) {
+    return status;
+  }
+  // Save attributes to attr managers.
+  for (uint32_t i = 0; i < attr_count; i++) {
+    this->attrMgr->set_attribute(idStr, attr_list[i]);
+  }
+
+  return SAI_STATUS_SUCCESS;
```

Comment:
This looks really similar to the function above. Can call into the other?

Reply:
i think it's different enough, i'm overloading the funcs in some cases, so it's actually calling different functions